### PR TITLE
feat(native-filters): Set default scope by filters' and charts' datasets 

### LINKED ISSUE #14977: [native-filters]upscope chart if the filter uses a different dataset?
When creating a new filter, in the scoping tab, should we uncheck the checkbox if the chart is using a different dataset than the filter uses. In this case, the default radio select should be set in `Apply to specific panels`. 
Currently with the new out of scope filter section, even if a filter is completely incompatible with a chart, user can still scope it and the incompatible filter would show in the scoped section(on the top open by default), which is somewhat confusing. 

Screen Shot 2021-06-03 at 3 56 29 PM

Screen Shot 2021-06-03 at 4 13 42 PM

Edge case: both available datasets have common column state.  

Screen Shot 2021-06-03 at 4 17 52 PM
Screen Shot 2021-06-03 at 4 19 14 PM

@villebro @kgabryje @michael-s-molina this proposed solution may not be ideal either, anything thoughts team?

### PR DESCRIPTION
**Pull request by kgabryje**
### SUMMARY
When new native filter was created, "Apply to all panels" scope was always selected by default. In the case of dashboards that have charts built on multiple datasets (e.g. `Slack Dashboard` from test data), filter that uses given dataset usually is incompatible and does not have any effect on charts that use a different dataset. This PR implements "smarter" selection of default scope - when user created a new filter, charts that use different dataset than the filter will be unselected. Changing dataset updates the default scoping tree, but only if the user didn't make any changes there - if the user changes selection in the scoping tree and then changes dataset, we don't update the tree automatically.
The feature takes effect only on new filters. Once a filter is saved, we don't apply default selection to the scopong tree upon editing the filter and its dataset.
Next to "incompatible" charts I added an icon with a tooltip that explains why those charts were not selected by default. I used `Info` icon for that purpose. Other icons that I considered were `Alert` and `Warning` , but since potentially a lot of those icons can appear next to each other, I didn't want to use something too "aggressive". I'd greatly appreciate design input/approval! CC @mihir174.

### BEFORE/AFTER SCREENSHOTS OR ANIMATED GIF
https://user-images.githubusercontent.com/15073128/122973217-47028c00-d391-11eb-8a1a-0c21d13ca50e.mov

### TESTING INSTRUCTIONS
0. Enable `DASHBOARD_NATIVE_FILTERS` feature flag
1. Create a dashboard with charts that use the same dataset.
2. Create a native filter and verify that "Apply to all panels" scope is selected by default
3. Create a dashboard with charts that use multiple datasets
4. Create a native filter, choose a dataset and verify that only the charts that use that dataset are in scope by default.
5. Change dataset, verify that scope tree got updated so that charts that use the new dataset are now in scope.
6. Modify the scope, change dataset and verify that the scope didn't change.
7. Save the filter, edit it, change its dataset, verify that the scoe didn't change.

### ADDITIONAL INFORMATION

- [x] Has associated issue: closes #14977 
- [ ] Changes UI
- [ ] Includes DB Migration (follow approval process in [SIP-59](https://github.com/apache/superset/issues/13351))
  - [ ] Migration is atomic, supports rollback & is backwards-compatible
  - [ ] Confirm DB migration upgrade and downgrade tested
  - [ ] Runtime estimates and downtime expectations provided
- [ ] Introduces new feature or API
- [ ] Removes existing feature or API

CC: @villebro @michael-s-molina @rusackas

### CODE REVIEW

File: superset-frontend/src/dashboard/components/nativeFilters/FiltersConfigModal/FiltersConfigForm/FilterScope/FilterScope.tsx [hunk 1]
```diff
@@ -56,11 +57,41 @@ const FilterScope: FC<FilterScopeProps> = ({
...
-  const initialScope = scope || getDefaultScopeValue(chartId);
-  const initialScoping = isScopingAll(initialScope, chartId)
-    ? Scoping.all
-    : Scoping.specific;
+  const [initialScope] = useState(
+    scope || getDefaultScopeValue(chartId, initiallyExcludedCharts),
+  );
+  const [initialScoping] = useState(
+    isScopingAll(initialScope, chartId) ? Scoping.all : Scoping.specific,
+  );
```
Comment (villebro) [inline superset-frontend/src/dashboard/components/nativeFilters/FiltersConfigModal/FiltersConfigForm/FilterScope/FilterScope.tsx:67]: This isn't directly related to this PR, but it confused me while reading this code: having two props called `initialScope` and `initialScoping` right next to each other feels very ambiguous. If changing this to something slightly more specific doesn't lead down a huge rabbit hole, I'd opt for renaming `Scoping` to `ScopingType` and consequently `initialScoping` to `initialScopingType`, perhaps even `initialScope` to `initialChartScope`.

File: superset-frontend/src/dashboard/components/nativeFilters/FiltersConfigModal/FiltersConfigForm/FiltersConfigForm.tsx [hunk 2]
```diff
@@ -580,6 +592,28 @@ const FiltersConfigForm = (
     setActiveFilterPanelKey(activeFilterPanelKey);
   }, [hasCheckedAdvancedControl]);
 
+  const initiallyExcluded = useMemo(() => {
```
Comment (villebro) [inline superset-frontend/src/dashboard/components/nativeFilters/FiltersConfigModal/FiltersConfigForm/FiltersConfigForm.tsx:595]: nit: `initiallyExcludedCharts`

**Review by villebro — COMMENTED** [review]
LGTM - a few minor readability improvement ideas. Also, should we reword this, especially if it will end up applying to charts referencing a different dataset, in which case the info icon could be helpful (maybe even stating how many charts are refefrencing a different dataset)?

image

File: superset-frontend/src/dashboard/components/nativeFilters/FiltersConfigModal/FiltersConfigForm/FilterScope/FilterScope.tsx (hunk 1 above)
Comment (kgabryje) [reply superset-frontend/src/dashboard/components/nativeFilters/FiltersConfigModal/FiltersConfigForm/FilterScope/FilterScope.tsx:67]: Totally agree, I feel like I spend more time trying to wrap my head around difference between those 2 than actual developing. I'll change that in this PR

**Review by villebro — APPROVED** [review]
Very nice! Extra points for the `useComponentDidUpdate` hook :D

**Comment by kgabryje** [comment]
@graceguo-supercat Would you like to verify that this PR works properly on Airbnb environments before it's merged?

File: superset-frontend/src/dashboard/components/nativeFilters/FiltersConfigModal/FiltersConfigForm/FilterScope/FilterScope.test.tsx [hunk 3]
```diff
@@ -115,6 +115,7 @@ describe('FilterScope', () => {
     fireEvent.click(screen.getByText('Scoping'));
     fireEvent.click(screen.getByLabelText('Apply to specific panels'));
 
+    screen.debug(undefined, 30000);
```
Comment (michael-s-molina) [inline superset-frontend/src/dashboard/components/nativeFilters/FiltersConfigModal/FiltersConfigForm/FilterScope/FilterScope.test.tsx:118]:
```suggestion
```

File: superset-frontend/src/common/hooks/useComponentDidUpdate/useComponentDidUpdate.ts [hunk 4]
```diff
@@ -0,0 +1,34 @@
...
+export const useComponentDidUpdate = (effect: EffectCallback) => {
+  const isMountedRef = useRef(false);
+  // eslint-disable-next-line react-hooks/exhaustive-deps
+  useEffect(() => {
+    if (isMountedRef.current) {
+      effect();
+    }
+  }, [effect]);
+
+  useEffect(() => {
```
Comment (michael-s-molina) [inline superset-frontend/src/common/hooks/useComponentDidUpdate/useComponentDidUpdate.ts:31]: Do we need this second `useEffect`? https://stackoverflow.com/questions/53179075/with-useeffect-how-can-i-skip-applying-an-effect-upon-the-initial-render

File: superset-frontend/src/dashboard/components/nativeFilters/FiltersConfigModal/FiltersConfigForm/FilterScope/FilterScope.test.tsx (hunk 3 above)
Comment (kgabryje) [reply superset-frontend/src/dashboard/components/nativeFilters/FiltersConfigModal/FiltersConfigForm/FilterScope/FilterScope.test.tsx]: 👍

File: superset-frontend/src/common/hooks/useComponentDidUpdate/useComponentDidUpdate.ts (hunk 4 above)
Comment (kgabryje) [reply superset-frontend/src/common/hooks/useComponentDidUpdate/useComponentDidUpdate.ts]: Good point, thanks!

**Review by michael-s-molina — APPROVED** [review]

**Pull request merged**